wordpress-mobile/WordPress-Android
# Add upload completion processor for VideoPress block

Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/5571

### Details

An media upload completion processor has been added for the new VideoPress block. The processor follows similar patterns to other media processors, particularly the processor for [the audio block](https://github.com/wordpress-mobile/WordPress-Android/blob/387847010b8884939d2e10f3ceb7f9a4b852114d/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/AudioBlockProcessor.kt), given that was the most recently added.

If a user navigates away from the editor while the VideoPress block is uploading, the processor will complete the upload and populate it with the expected content. In order to achieve, it's necessary to build a VideoPress URL based on the block's various settings. You can see how this is achieved in Jetpack [here](https://github.com/Automattic/jetpack/blob/b1b826ab38690c5fad18789301ac81297a458878/projects/packages/videopress/src/client/lib/url/index.ts#L19-L67), and a similar pattern has been followed in this PR.

The iOS equivalent of this PR can be found at https://github.com/wordpress-mobile/WordPress-iOS/pull/20380

### Testing

1. Add a VideoPress block to an existing or new post.
2. Select the **choose from device** option and upload a video.
3. While there’s an ongoing upload, close the post with publishing changes.
4. Verify that you see the upload progress in the post summary.
5. Re-open the post and confirm the upload has completed.
6. Preview the post and observe that the video can be played.

### Default markup examples

**Example default markup before upload:**

```

<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"></figure>

```

**Example default markup after upload:**

```

<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"><div class="jetpack-videopress-player__wrapper">https://videopress.com/v/AbCdE?resizeToParent=true&amp;cover=true&amp;preloadContent=metadata&amp;useAverageColor=true</div></figure>

```

### Markup with attributes examples

**Example markup before upload:**

```

<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"></figure>

```

**Example markup after upload:**

```

<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"><div class="jetpack-videopress-player__wrapper">https://videopress.com/v/AbCdE?resizeToParent=true&amp;cover=true&amp;autoPlay=true&amp;controls=false&amp;loop=true&amp;muted=true&amp;persistVolume=false&amp;playsinline=true&amp;posterUrl=https%3A%2F%2Ftest.files.wordpress.com%2F2022%2F02%2F265-5000x5000-1.jpeg&amp;preloadContent=none&amp;sbc=%23abb8c3&amp;sbpc=%239b51e0&amp;sblc=%23cf2e2e</div></figure>

```

## Regression Notes
1. Potential unintended areas of impact

This change might impact other upload processors, with the video processor being a specific concern due to the possibility of the similar block names being matched.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I manually tested uploading media to both the video and VideoPress blocks. I also relied on the existing automated tests for each of the processors to flag any issues and added new tests for the VideoPress block.

3. What automated tests I added (or what prevented me from doing so)

* https://github.com/wordpress-mobile/WordPress-Android/pull/18187/commits/91cc0bdd7f5352bb0c04e51ef38d4ecd3bdbb843: Test to check the video block processor doesn't change anything in the VideoPress block.
* https://github.com/wordpress-mobile/WordPress-Android/pull/18187/commits/ec71ddbe594d64720f1dc371367204b9abba4af6: Test to check the VideoPress block processor doesn't change anything in the video block.
* https://github.com/wordpress-mobile/WordPress-Android/pull/18187/commits/f36eb72bd561f93fed04a66340a0c9ae3b2b8dd0: Test to ensure processor correctly converts VideoPress block with default attributes. I followed the same pattern set out [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/2ddbacb28a54ef608e1250256c4b05381d3662d4/WordPress/WordPressTest/GutenbergVideoPressUploadProcessorTests.swift#L7-L30) for iOS. 
* https://github.com/wordpress-mobile/WordPress-Android/pull/18187/commits/4083c4d233659ae46c818d890cb24fe8cc31625b: Test to ensure processor correctly converts VideoPress block with non-default attributes. I followed the same pattern set out [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/2ddbacb28a54ef608e1250256c4b05381d3662d4/WordPress/WordPressTest/GutenbergVideoPressUploadProcessorTests.swift#L32-L55) for iOS.

<hr />

PR submission checklist:
